src/clj/fluree/db/db/json_ld.cljc

```diff
@@ -474,60 +477,84 @@
+  [assert? db sid pid t value]
   (let [v-maps (util/sequential value)]
     (mapcat (fn [v-map]
               (if (list-value? v-map)
                 (let [list-vals (:list v-map)]
                   (into []
                         (comp (map add-list-meta)
-                              (map (partial assert-value-map db sid pid t)))
+                              (map (partial value-map->flake assert? db sid pid t)))
                         list-vals))
-                [(assert-value-map db sid pid t v-map)]))
+                [(value-map->flake assert? db sid pid t v-map)]))
             v-maps)))
 
-(defn- get-type-assertions
-  [db t sid type]
-  (if type
-    (loop [[type-item & r] type
-           acc []]
-      (if type-item
-        (let [type-id (iri/encode-iri db type-item)]
-          (recur r (conj acc (flake/create sid const/$rdf:type type-id const/$xsd:anyURI t true nil))))
-        acc))
-    []))
-
-(defn assert-node
-  [db t node]
-  (log/trace "assert-node:" node)
+(defn- get-type-flakes
+  [assert? db t sid type]
+  (into []
+        (map (fn [type-item]
+               (let [type-sid (iri/encode-iri db type-item)]
+                 (flake/create sid const/$rdf:type type-sid
+                               const/$xsd:anyURI t assert? nil))))
+        type))
+
+(defn encode-existing-iri
+  "Throws if IRI doesn't exist"
+  [db id]
+  (or (iri/encode-iri db id)
+      (throw (ex-info (str "IRI does not exist: " id)
+                      {:status 400, :error :db/invalid-commit}))))
+
+(defn node->flakes
+  [assert? db t node]
+  (log/trace "node->flakes:" node "assert?" assert?)
   (let [{:keys [id type]} node
-        sid             (iri/encode-iri db id)
+        sid             (if assert?
+                          (iri/encode-iri db id)
+                          (try*
+                            (encode-existing-iri db id)
+                            (catch* e
+                              (throw
+                               (ex-info
+                                (str "Retraction specifies an invalid IRI at db t value: "
+                                     t ": " (ex-message e))
+                                {:status 400, :error :db/invalid-commit})))))
         type-assertions (if (seq type)
-                          (get-type-assertions db t sid type)
+                          (get-type-flakes assert? db t sid type)
                           [])]
     (into type-assertions
-          (comp (filter (fn [node-entry]
-                          (not (-> node-entry key keyword?))))
-                (mapcat (fn [[prop value]]
-                          (let [pid (iri/encode-iri db prop)]
-                            (assert-property db sid pid t value)))))
+          (comp (remove #(-> % key keyword?))
+                (mapcat
+                 (fn [[prop value]]
+                   (let [pid (if assert?
+                               (iri/encode-iri db prop)
+                               (try*
```

Comment:
Instead of doing this messy exception handling here, could we just let the `nil` from `iri/encode-iri` trickle down to `value-map->flake` and throw if `pid` or `sid` are `nil` when we actually are about to call `flake/create`?

Also, I think we should update the error value on the exception, this has nothing to do with committing and I think that can only cause confusion (though I realize you're not the one who wrote the error initially).

Reply:
> Instead of doing this messy exception handling here, could we just let the `nil` from `iri/encode-iri` trickle down to `value-map->flake` and throw if `pid` or `sid` are `nil` when we actually are about to call `flake/create`?

I like keeping the exception handling more explicit and closely and directly connected to the problem it's dealing with. I hate it when I run into code handling "maybe this is `nil`?" and I then have to go figure out why / when it could be nil. This doesn't seem all that messy to me compared to that.

> Also, I think we should update the error value on the exception, this has nothing to do with committing and I think that can only cause confusion (though I realize you're not the one who wrote the error initially).

These fns are all in support of `merge-commit`, so it seems commit-related to me, unless I'm misunderstanding?

Reply:
Fair enough, regarding letting nils by. 

Do we need `encode-existing-iri` though? We could just use an `or` instead of `try*` for the same effect in fewer lines.

As for whether this is commit related, this is all actually downstream of `stage`, and `merge-commit` just happens to reuse that machinery. If there's any problem it'll only surface during a call to `stage`. Commits should only have known-good data.

Reply:
> Fair enough, regarding letting nils by.
> 
> Do we need `encode-existing-iri` though? We could just use an `or` instead of `try*` for the same effect in fewer lines.

I like giving things like that intention-revealing names. It wasn't immediately obvious to me at first glance why this was different for retracts until I thought about it for a second. But I'm not super wedded to it in this case, so if you don't think it's worth it I can go back to an `or`.

> As for whether this is commit related, this is all actually downstream of `stage`, and `merge-commit` just happens to reuse that machinery. If there's any problem it'll only surface during a call to `stage`. Commits should only have known-good data.

OK I can tweak the error message.

Reply:
Oh, wait. I now realize I don't think I follow your suggested change to the error. Can you elaborate?

Reply:
I just found the pattern of throwing an exception then immediately catching it and throwing a different exception to be a little heavy for this case, the exception should provide enough context about why it's necessary.

Reply:
Oh, I see. I did that because in the `encode-existing-iri` context it shouldn't assume things that the original error messages contained (see the deleted retract code). But I didn't want to lose information that those error messages had in case anyone was relying on them. So I throw an error from `encode-existing-iri` only containing what it knows went wrong, but then added in the additional context from the original errors in the calling catches.

Reply:
Hold on, this isn't downstream of `stage` at all - I was wrong. I thought they shared the same mechanism but it looks like it's different. So it looks like the error will only show up while attempting to load a ledger.

Here's kind of what I'm thinking:
```
(if assert?
  (iri/encode-iri db prop)
  (or (iri/encode-iri db prop)
        (throw (ex-info "Cannot retract IRI with unknown namespace." 
                                    {:status 400 :error :db/invalid-retraction :iri prop}))))
```